wagtail/core/collectors.py

```diff
@@ -0,0 +1,704 @@
+            else:
+                use = cls(obj, parent=parent, on_delete=on_delete)
+                if use in exclude:
+                    if use not in originals:
+                        for other_use in exclude:
+                            if use == other_use:
+                                main_use = other_use
+                                break
+                else:
+                    exclude.add(use)
+                    yield use
+                    main_use = use
+
+    def __hash__(self):
+        return hash(self.key)
+
+    def __eq__(self, other):
+        if isinstance(other, Use):
+            return self.key == other.key
+        return self.key == get_obj_base_key(other)
+
+    @property
+    def is_root(self):
+        return self.depth == 0
+
+    @property
+    def is_protected(self):
+        return self.on_delete == PROTECT
+
+    def is_hidden(self):
+        if self.is_protected:
+            return False
+        if isinstance(self.object, TaggedItemBase):
+            return True
+        if hasattr(self.object, 'is_shown_in_uses') \
+                and not self.object.is_shown_in_uses():
+            return True
+        return any(isinstance(field, (ParentalKey, ParentalManyToManyField))
+                   for field in self.object._meta.fields)
+
+    def get_on_delete_data(self):
+        if self.on_delete == CASCADE:
+            return 'serious', _('Will also be deleted')
+        if self.on_delete == PROTECT:
+            return 'primary', _('Prevents deletion')
+        if self.on_delete == SET_NULL:
+            if self.field is None:
+                return '', _('Field will be emptied')
+            return '', _('Field “%s” will be emptied') % self.field.verbose_name
+        if self.on_delete == SET_DEFAULT:
+            return 'primary', _('Default object will be set')
+        if self.on_delete == DO_NOTHING:
+            return '', _('Nothing will happen')
+        return 'primary', _('Another object will be set')
+
+    def get_on_delete_html(self):
+        return format_html('<span class="status-tag {}">{}</span>',
+                           *self.get_on_delete_data())
+
+    @property
+    def model_name(self):
+        return self.object._meta.verbose_name
+
+    def get_edit_link(self):
+        from wagtail.admin.viewsets import viewsets
+
+        if hasattr(self.object, 'get_edit_url'):
+            url = self.object.get_edit_url()
+        else:
+            viewset = viewsets.get_for_model(self.object._meta.model)
+            if viewset is None:
+                return str(self.object)
+            url = reverse(viewset.get_url_name('edit'), args=(self.object.pk,))
+        return format_html('<a href="{}">{}</a>', url, self.object)
+
+    def get_html(self):
+        html = self.get_edit_link()
+        if self.is_root:
+            return html
+        return format_html('{}<i class="icon icon-arrow-right"></i> {}',
+                           (' ' * self.depth * 8), html)
```

Comment:
Oh, this part got lost somehow, maybe during a merge: the space should be a non-breaking space, I’m sure I originally put it. @gasman please include it again, but with an explicit unicode code: `'\u00A0'`